### PR TITLE
RFC: Fix ffmpeg leaks

### PR DESCRIPTION
**Pull request by OznOg**
### What does this PR do?
Try to fix leaks in ffmpeg
please let me know it I missed some of them

### Closes Issue(s)
probably #444

### Motivation
Prevent resources leaks

### CODE REVIEW

**Comment by OznOg** [comment]
Do we want to keeep support on old libav or not?

File: game/ffmpeg.cc [hunk 1]
```diff
@@ -302,10 +315,11 @@ class FfmpegError: public std::runtime_error {
 
 void FFmpeg::decodePacket() {
 #if LIBAVCODEC_VERSION_INT >= (AV_VERSION_INT(57, 37, 0))
-	AVPacket pkt;
```
Comment (Lord-Kamina) [inline game/ffmpeg.cc]: Is this intended? Why'd you put the declaration inside the loop?

**Comment by Lord-Kamina** [comment]
> Do we want to keeep support on old libav or not?



> Do we want to keeep support on old libav or not?

I don't really know how old that libav is, can you elaborate?

**Comment by Lord-Kamina** [comment]
Actually, I checked... we've got compatibility code there going all the way back to FFmpeg 1.0; that's fucking insane. I propose we keep compatbility up to FFmpeg 3 (Major version number 57 for most components)

**Comment by Tronic** [comment]
Is libav still used somewhere? I never followed how that played out after Debian began shipping libav in its ffmpeg packages and all that. If all major distributions are back to ffmpeg already, I don't see point in supporting both. Same goes for old versions because ffmpeg tends to be rapidly updated in distributions and its APIs are frequently deprecated.

File: game/ffmpeg.cc (hunk 1 above)
Comment (OznOg) [reply game/ffmpeg.cc]: Yes this is intendended for now, I need more tests anyway. Don't bother too much with reviews for now, the purpose of the PR was mainly to know your opinion on the legacy AV lib we want to support. I'll let you know when the code is ready for reviews.

**Comment by OznOg** [comment]
IMHO, the less #ifdef we have the less problems we have, nevertheless, what is the oldest distro begin supported? is ubuntu 16.04 considered outdated for us or not?

File: game/ffmpeg.cc (hunk 1 above)
Comment (Baklap4) [reply game/ffmpeg.cc]: Setting this as 'resolved'

**Comment by Baklap4** [comment]
I suppose ubuntu 16.04 is our oldest target for now.
Semaphore has been updated to also include 18.04 builds

on ubuntu 16.04 libavcodec-dev version 7:3.4.2-2build2ppa-xenial1 is used
which depends on libavutil-ffmpeg54

**Comment by Lord-Kamina** [comment]
> Is libav still used somewhere? I never followed how that played out after Debian began shipping libav in its ffmpeg packages and all that. If all major distributions are back to ffmpeg already, I don't see point in supporting both. Same goes for old versions because ffmpeg tends to be rapidly updated in distributions and its APIs are frequently deprecated.

Sorry, I tend to use "libav" and "ffmpeg" interchangeably when I write.

**Comment by OznOg** [comment]
> Is libav still used somewhere? I never followed how that played out after Debian began shipping libav in its ffmpeg packages and all that. If all major distributions are back to ffmpeg already, I don't see point in supporting both. Same goes for old versions because ffmpeg tends to be rapidly updated in distributions and its APIs are frequently deprecated.

yes, my bad, I meant ffmpeg, but flags actually are LIBAV_CODEC and so on...

**Comment by Lord-Kamina** [comment]
> LGTM @Lord-Kamina can you also take a look?

I'm not sure I understand the implications of all the changes to be honest but looks ok.

**Comment by OznOg** [comment]
agree that it is quite a lot of changes, but can be summed u like this:
* Remove legacy ffmpeg compat
* add unique_ptr to implement RAII
* remove dead code (in while loops)

I used https://www.ffmpeg.org/doxygen/3.4/filtering_video_8c-example.html#a90 as "reference implmentation"

File: game/ffmpeg.cc [hunk 2]
```diff
@@ -170,85 +159,129 @@ FFmpeg::FFmpeg(fs::path const& _filename, unsigned int rate):
...
+				throw FfmpegError(ret);
+			}
+
+			if (pkt.stream_index == m_streamId) {
+
+				if (audioQueue.wantSeek()) m_seekTarget = 0.0;
+				if (m_seekTarget == m_seekTarget) seek_internal();
+				decodePacket(pkt);
+				errors = 0;
+			}
```
Comment (Tronic) [inline game/ffmpeg.cc]: Does this (still) support seeking back from near to the end of a song, once the decoder has already reached EOF?

Comment (Baklap4) [reply game/ffmpeg.cc:284]: @OznOg can you test/check/verify this :)?

**Comment by OznOg** [comment]
Well as said in previous comment, I tested the use case I understood without notifying any difference with master, but I would need any power user to confirm...

File: game/ffmpeg.cc (hunk 2 above)
Comment (Baklap4) [reply game/ffmpeg.cc:284]: Running a song almost till the end and then rewind still works!
There is still a bug from a few years ago https://github.com/performous/performous/issues/33
which sadly still applies.

**Comment by Baklap4** [comment]
@OznOg can you check if you might be able to fix #33 aswell on this branch?

**Pull request merged**